src/compile-cache.js

```diff
@@ -0,0 +1,141 @@
+}
+
+var cacheDirectory = null
+
+Object.keys(COMPILERS).forEach(function (extension) {
+  var compiler = COMPILERS[extension]
+
+  Object.defineProperty(require.extensions, extension, {
+    enumerable: true,
+    writable: false,
+    value: function (module, filePath) {
+      var code = compileFileAtPath(compiler, filePath, extension)
+      return module._compile(code, filePath)
+    }
+  })
+})
+
+exports.setAtomHomeDirectory = function (atomHome) {
+  var cacheDir = path.join(atomHome, 'compile-cache')
+  if (process.env.USER === 'root' && process.env.SUDO_USER && process.env.SUDO_USER !== process.env.USER) {
+    cacheDir = path.join(cacheDirectory, 'root')
+  }
+  this.setCacheDirectory(cacheDir)
+}
+
+exports.setCacheDirectory = function (directory) {
+  cacheDirectory = directory
+}
+
+exports.getCacheDirectory = function () {
+  return cacheDirectory
+}
+
+exports.addPathToCache = function (filePath, atomHome) {
+  this.setAtomHomeDirectory(atomHome)
+  var extension = path.extname(filePath)
+
+  if (extension === '.cson') {
+    CSON = CSON || require('season')
+    CSON.readFileSync(filePath)
+  } else {
+    var compiler = COMPILERS[extension]
+    if (compiler) {
+      compileFileAtPath(compiler, filePath, extension)
+    }
+  }
+}
+
+function compileFileAtPath (compiler, filePath, extension) {
+  var sourceCode = fs.readFileSync(filePath, 'utf8')
+  if (compiler.shouldCompile(sourceCode, filePath)) {
+    var cachePath = compiler.getCachePath(sourceCode, filePath)
+    var compiledCode = readCachedJavascript(cachePath)
+    if (compiledCode == null) {
+      compiledCode = addSourceURL(compiler.compile(sourceCode, filePath), filePath)
+      writeCachedJavascript(cachePath, compiledCode)
+    }
+    return compiledCode
+  }
+  return sourceCode
+}
+
+function readCachedJavascript (relativeCachePath) {
+  var cachePath = path.join(cacheDirectory, relativeCachePath)
+  if (fs.isFileSync(cachePath)) {
+    try {
+      return fs.readFileSync(cachePath, 'utf8')
+    } catch (error) {}
+  }
+  return null
+}
+
+function writeCachedJavascript (relativeCachePath, code) {
+  var cachePath = path.join(cacheDirectory, relativeCachePath)
+  fs.writeFileSync(cachePath, code, 'utf8')
+}
+
+function addSourceURL (jsCode, filePath) {
+  if (process.platform === 'win32')
+    filePath = '/' + path.resolve(filePath).replace(/\\/g, '/')
+  return jsCode + '\n' + '//# sourceURL=' + encodeURI(filePath) + '\n'
+}
+
+var INLINE_SOURCE_MAP_REGEXP = /\/\/[#@]\s*sourceMappingURL=([^'"\n]+)\s*$/mg
+
+require('source-map-support').install({
+  handleUncaughtExceptions: false,
+
+  // Most of this logic is the same as the default implementation in the
+  // source-map-support module, but we've overridden it to read the javascript
+  // code from our cache directory.
+  retrieveSourceMap: function (filePath) {
+    if (!fs.isFileSync(filePath)) {
+      return null
+    }
+
+    var sourceCode = fs.readFileSync(filePath, 'utf8')
```

Comment:
Ok, I'm just gonna let it behave like there is no source map, in the event that there's an IO error retrieving the source map.